# Exploring SnapShot Results Capabilities with `py_dss_toolkit`: A Hands-on Example
This notebook demonstrates how to compile and solve a DSS model, and then retrieve key results, such as circuit summaries, voltages, powers, and currents.

In [72]:
import py_dss_interface
from py_dss_toolkit import dss_tools

Define the path to the DSS file

In [73]:
dss_file = r"C:\PauloRadatz\GitHub\py_dss_toolkit\examples\feeders\123Bus\IEEE123Master.dss"

Create an instance of DSS

In [74]:
dss = py_dss_interface.DSS()

Connect DSS instance to the dss_tools

In [75]:
dss_tools.update_dss(dss)

Compile the DSS model from the specified file, load bus coordinates for visualization

In [76]:
dss.text(f"compile [{dss_file}]")
dss.text(f"buscoords buscoords.dat")

''

Solve the power flow using the `solve_snapshot` method from dss_tools with customized parameters. For example, we set the maximum control iterations to 20. There are other parameters you can change.

In [77]:
dss_tools.simulation.solve_snapshot(max_control_iter=20)

Retrieve the circuit summary as a DataFrame

In [78]:
circuit_summary_df = dss_tools.results.summary_df
circuit_summary_df

Results
P feeder (kW)     3615.241922
Q feeder (kvar)   1311.514987
P losses (kW)       95.976865
Q losses (kvar)    192.504095
max voltage (pu)     1.049961
min voltage (pu)     0.979211

Get the magnitudes and angles of voltages (line-neutral) for each node

In [79]:
mag_voltage_ln_nodes_df = dss_tools.results.voltage_ln_nodes[0]
angle_voltage_ln_nodes_df = dss_tools.results.voltage_ln_nodes[1]

mag_voltage_ln_nodes_df.head(), angle_voltage_ln_nodes_df.head()

(         node1     node2     node3
 150   0.999990  0.999994  0.999993
 150r  1.037486  1.037492  1.037490
 149   1.037486  1.037492  1.037490
 1     1.024960  1.035118  1.028590
 2          NaN  1.034896       NaN,
          node1       node2       node3
 150  -0.001453 -120.000956  119.998818
 150r -0.001954 -120.001285  119.998410
 149  -0.001949 -120.001282  119.998414
 1    -0.648331 -120.316266  119.619434
 2          NaN -120.320463         NaN)

Retrieve the active and reactive powers (P and Q) for each element in the system

In [80]:
p_elements_df = dss_tools.results.powers_elements[0]
q_elements_df = dss_tools.results.powers_elements[1]

p_elements_df.head(), q_elements_df.head()

(                   Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  \
 transformer.reg1a  1463.277554   962.106773  1189.857596          0.0   
 line.l115          1463.277006   962.106543  1189.857248          NaN   
 line.l1                    NaN    20.003555          NaN          NaN   
 line.l2                    NaN          NaN   102.853919          NaN   
 line.l3            1411.964368   941.731825  1079.372647          NaN   
 
                    Terminal2.1  Terminal2.2  Terminal2.3  Terminal2.0  
 transformer.reg1a -1463.277405  -962.106710 -1189.857502          0.0  
 line.l115         -1451.964535  -961.735379 -1182.226566          NaN  
 line.l1                    NaN   -19.999990          NaN          NaN  
 line.l2                    NaN          NaN  -102.717428          NaN  
 line.l3           -1403.458648  -941.604701 -1074.807433          NaN  ,
                    Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  \
 transformer.reg1a   578.377078   336.317

Get the magnitudes and angles of currents for each element in the system

In [81]:
mag_current_df = dss_tools.results.currents_elements[0]
angle_current_df = dss_tools.results.currents_elements[1]

mag_current_df.head(), angle_current_df.head()

(                   Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  \
 transformer.reg1a   655.119865   424.352923   522.235105   183.173775   
 line.l115           631.440834   409.014866   503.359137          NaN   
 line.l1                    NaN     8.995957          NaN          NaN   
 line.l2                    NaN          NaN    46.569548          NaN   
 line.l3             613.365422   400.100368   457.360338          NaN   
 
                    Terminal2.1  Terminal2.2  Terminal2.3  Terminal2.0  
 transformer.reg1a   631.440834   409.014866   503.359137   176.553037  
 line.l115           631.441479   409.015224   503.359437          NaN  
 line.l1                    NaN     8.996124          NaN          NaN  
 line.l2                    NaN          NaN    46.569785          NaN  
 line.l3             613.365893   400.100628   457.360543          NaN  ,
                    Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  \
 transformer.reg1a   -21.568426  -139.268

Get the loading of the elements.

In [82]:
current_loading_df = dss_tools.results.current_loading_percent
current_loading_df.head()

Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  \
transformer.reg1a    62.938076    40.768046    50.171693    17.597703   
line.l115           105.240139    68.169144    83.893190          NaN   
line.l1                    NaN     1.499326          NaN          NaN   
line.l2                    NaN          NaN     7.761591          NaN   
line.l3             102.227570    66.683395    76.226723          NaN   

                   Terminal2.1  Terminal2.2  Terminal2.3  Terminal2.0  
transformer.reg1a    60.663206    39.294502    48.358258    16.961642  
line.l115           105.240247    68.169204    83.893239          NaN  
line.l1                    NaN     1.499354          NaN          NaN  
line.l2                    NaN          NaN     7.761631          NaN  
line.l3             102.227649    66.683438    76.226757          NaN

Get the loading of the elements using the emergency amps.

In [83]:
dss_tools.results.set_violation_current_limit_type("emerg_amps")
current_loading_df = dss_tools.results.current_loading_percent
current_loading_df.head()

Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  \
transformer.reg1a    62.938076    40.768046    50.171693    17.597703   
line.l115           105.240139    68.169144    83.893190          NaN   
line.l1                    NaN     1.499326          NaN          NaN   
line.l2                    NaN          NaN     7.761591          NaN   
line.l3             102.227570    66.683395    76.226723          NaN   

                   Terminal2.1  Terminal2.2  Terminal2.3  Terminal2.0  
transformer.reg1a    60.663206    39.294502    48.358258    16.961642  
line.l115           105.240247    68.169204    83.893239          NaN  
line.l1                    NaN     1.499354          NaN          NaN  
line.l2                    NaN          NaN     7.761631          NaN  
line.l3             102.227649    66.683438    76.226757          NaN

Get nodal voltage violations, voltages outside the default range of 0.95 and 1.05 pu

In [84]:
violations_mag_voltage_ln_nodes_df = dss_tools.results.violation_voltage_ln_nodes
# Undervoltage
violations_mag_voltage_ln_nodes_df[0].head()
# Overvoltage
violations_mag_voltage_ln_nodes_df[1].head()

Empty DataFrame
Columns: [node1, node2, node3]
Index: []

Change the range to 0.98 and 1.04 pu.

In [85]:
dss_tools.results.set_violation_voltage_ln_limits(v_min_pu=0.98, v_max_pu=1.04)
violations_mag_voltage_ln_nodes_df = dss_tools.results.violation_voltage_ln_nodes

In [86]:
# Undervoltage
violations_mag_voltage_ln_nodes_df[0].head()

node1     node2     node3
64  0.979896  1.016154  0.993659
65  0.979211  1.015822  0.990654
66  0.979450  1.016063  0.989139

In [87]:
# Overvoltage
violations_mag_voltage_ln_nodes_df[1].head()

node1     node2     node3
67  1.041080  1.044678  1.034176
72  1.041488  1.043832  1.033952
97  1.040082  1.044185  1.033443
76  1.041406  1.043345  1.034566
77  1.042572  1.044538  1.035443

Get elements with thermal violations

In [88]:
dss_tools.results.get_violation_current_limit_type()

'emerg_amps'

In [89]:
# Back to normal amps
dss_tools.results.set_violation_current_limit_type("norm_amps")
violations_mag_current_df = dss_tools.results.violation_currents_elements
violations_mag_current_df.head()

Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  Terminal2.1  \
line.l115   157.860208   102.253716   125.839784          NaN   157.860370   
line.l3     153.341356   100.025092   114.340085          NaN   153.341473   
line.l7     151.063751   100.025157   114.340136          NaN   151.063828   
line.l10    139.515774    97.793473   114.340169          NaN   139.515886   
line.sw1    157.860208   102.253716   125.839784          NaN   157.860208   

           Terminal2.2  Terminal2.3  Terminal2.0  
line.l115   102.253806   125.839859          NaN  
line.l3     100.025157   114.340136          NaN  
line.l7     100.025200   114.340169          NaN  
line.l10     97.793535   114.340218          NaN  
line.sw1    102.253716   125.839784          NaN

Get thermal violations using the emergency amps.

In [90]:
dss_tools.results.set_violation_current_limit_type("emerg_amps")
violations_mag_current_df = dss_tools.results.violation_currents_elements
violations_mag_current_df.head()

Terminal1.1  Terminal1.2  Terminal1.3  Terminal1.0  Terminal2.1  \
line.l115   105.240139    68.169144    83.893190          NaN   105.240247   
line.l3     102.227570    66.683395    76.226723          NaN   102.227649   
line.l7     100.709168    66.683438    76.226757          NaN   100.709219   
line.sw1    105.240139    68.169144    83.893190          NaN   105.240139   

           Terminal2.2  Terminal2.3  Terminal2.0  
line.l115    68.169204    83.893239          NaN  
line.l3      66.683438    76.226757          NaN  
line.l7      66.683466    76.226779          NaN  
line.sw1     68.169144    83.893190          NaN

### Conclusion
In this notebook, we explored the SnapShot Results capabilities of py_dss_toolkit.
We compiled and solved a DSS model, then retrieved key metrics such as circuit summary data, voltages, powers, and currents.
If you have suggestions or ideas on what could be added to this tool, please don't hesitate to reach out. Your feedback is invaluable in helping improve this tool's functionality.